### PR TITLE
Document webrender_api::image

### PR DESCRIPTION
**Pull request by bholley**
Added docs to most things, stubbed out a few, and added deny(missing_docs).

r? @gw3583 or @kvark

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/3072)

### CODE REVIEW

File: webrender_api/src/image.rs [hunk 1]
```diff
@@ -74,17 +114,35 @@ impl ImageFormat {
...
+    /// Format of the image data.
     pub format: ImageFormat,
+    /// Width and length of the image data, in pixels.
     pub size: DeviceUintSize,
+    /// The number of bytes from the start of one row to the next. Lazily
+    /// computed via compute_stride() and cached here when the image is
+    /// tile.
+    ///
+    /// FIXME(bholley): The computation of this looks pretty cheap. Do we have
+    /// evidence that this is hot enough that it needs caching?
```
Comment (emilio) [inline webrender_api/src/image.rs]: This doesn't look like a cache, it's an override for when the stride is not width times bpp.

File: webrender_api/src/image.rs [hunk 2]
```diff
@@ -222,44 +298,64 @@ pub trait BlobImageHandler: Send {
 
 /// A group of rasterization requests to execute synchronously on the scene builder thread.
 pub trait AsyncBlobImageRasterizer : Send {
+    /// Rasterize the requests.
     fn rasterize(&mut self, requests: &[BlobImageParams]) -> Vec<(BlobImageRequest, BlobImageResult)>;
 }
 
 
+/// FIXME: document
```
Comment (emilio) [inline webrender_api/src/image.rs]: Maybe just allow() here to avoid having the comment below?

File: webrender_api/src/image.rs (hunk 1 above)
Comment (nical) [reply webrender_api/src/image.rs]: Yep, what Emilio said. Different source of images have different alignment constraints for rows, so the stride isn't always equal to width * bpp.

File: webrender_api/src/image.rs [hunk 3]
```diff
@@ -222,44 +298,64 @@ pub trait BlobImageHandler: Send {
...
 #[derive(Copy, Clone, Debug)]
 pub struct BlobImageParams {
+    /// FIXME: document
     pub request: BlobImageRequest,
+    /// FIXME: document
     pub descriptor: BlobImageDescriptor,
+    /// FIXME: document
```
Comment (nical) [inline webrender_api/src/image.rs]: /// An optional sub-rectangle of the image to avoid re-rasterizing the entire image when only a portion is updated.
///
/// If set to `None` the entire image is rasterized.

File: webrender_api/src/image.rs [hunk 4]
```diff
@@ -222,44 +298,64 @@ pub trait BlobImageHandler: Send {
...
 pub type BlobImageData = Vec<u8>;
 
+/// Result type for blob raserization.
 pub type BlobImageResult = Result<RasterizedBlobImage, BlobImageError>;
 
+/// Metadata (but not storage) for a blob image.
 #[repr(C)]
 #[derive(Copy, Clone, Debug)]
 pub struct BlobImageDescriptor {
+    /// Bounding rectangle of the image, in device pixels.
```
Comment (nical) [inline webrender_api/src/image.rs]: I'd rephrase this as:

/// Size in device pixels of the blob's output image.

File: webrender_api/src/image.rs [hunk 5]
```diff
@@ -222,44 +298,64 @@ pub trait BlobImageHandler: Send {
...
     pub size: DeviceUintSize,
+    /// FIXME - Document me.
     pub offset: DevicePoint,
```
Comment (nical) [inline webrender_api/src/image.rs]: Huh! it's unused. We should probably remove it.

File: webrender_api/src/image.rs [hunk 6]
```diff
@@ -222,44 +298,64 @@ pub trait BlobImageHandler: Send {
...
 pub enum BlobImageError {
+    /// Out of memory.
     Oom,
-    InvalidKey,
-    InvalidData,
+    /// Other failure, embedding-specified.
     Other(String),
 }
 
+/// FIXME: document
```
Comment (nical) [inline webrender_api/src/image.rs]: A key identifying blob image rasterization work requested to the blob image rasterizer.

File: webrender_api/src/image.rs [hunk 7]
```diff
@@ -222,44 +298,64 @@ pub trait BlobImageHandler: Send {
...
 #[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
 pub struct BlobImageRequest {
+    /// Unique handle to the image.
     pub key: ImageKey,
+    /// Tiling offset.
+    /// FIXME - document tiling better.
```
Comment (nical) [inline webrender_api/src/image.rs]: /// Tile offset in number of tiles, if applicable.
///
/// If set to `None` it means the image is not tiled.

File: webrender_api/src/image.rs [hunk 8]
```diff
@@ -74,17 +114,35 @@ impl ImageFormat {
...
+    /// Width and length of the image data, in pixels.
     pub size: DeviceUintSize,
+    /// The number of bytes from the start of one row to the next. Lazily
+    /// computed via compute_stride() and cached here when the image is
+    /// tile.
+    ///
+    /// FIXME(bholley): The computation of this looks pretty cheap. Do we have
+    /// evidence that this is hot enough that it needs caching?
     pub stride: Option<u32>,
+    /// Byte offset used for tiling. FIXME: Better documentation.
```
Comment (nical) [inline webrender_api/src/image.rs]: /// Offset in bytes of the first pixel in the buffer that holds the image.

File: webrender_api/src/image.rs (hunk 5 above)
Comment (bholley) [reply webrender_api/src/image.rs]: It's used by wrench.
